homeassistant/components/trest_solar/coordinator.py

```diff
@@ -0,0 +1,63 @@
+from __future__ import annotations  # noqa: D100
+
+from datetime import timedelta
+import logging
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import REFRESH_INTERVAL
+from .models.solar_history import SolarHistory
+from .services.trest_solar_service import CloudSolarTrestService
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TrestDataCoordinator(DataUpdateCoordinator[SolarHistory]):
+    """Trest coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        trest_solar_service: CloudSolarTrestService,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            logger,
+            name="Trest Solar Controller",
+            update_interval=timedelta(seconds=REFRESH_INTERVAL),
+        )
+        self.trest_solar_service = trest_solar_service
+
+    async def _async_update_data(self) -> SolarHistory:
+        """Fetch data from Trest Cloud."""
+
+        _LOGGER.info("Requesting Solar History data from Cloud Trest")
+        response = await self.trest_solar_service.get_latest_solar_history_async()
```

Comment:
Yes, it can, how would I handle this best? If the user is unauthorized, timeout or the service is down, then it would throw an exception, do you have any tips?

Reply:
You would raise `ConfigEntryAuthFailed` on auth errors to trigger a reauthentication flow (but add that in a separate PR later, for now you'd raise `ConfigEntryError`) and `UpdateFailed` on generic update errors (`from homeassistant.helpers.update_coordinator import UpdateFailed`)